### PR TITLE
[ufc] add null operator and more fixes

### PR DESCRIPTION
**Pull request by schmit**
---
labels: mergeable
---
[//]: #  (Link to the issue corresponding to this chunk of work)
Fixes: #__issue__

## Motivation and Context
[//]: #  (Why is this change required? What problem does it solve?)

- Adding the IS_NULL operator
- Type check return value
- Fix date logic

## Description
[//]: # (Describe your changes in detail)

## How has this been tested?
[//]: # (Please describe in detail how you tested your changes)

[//]: # (OPTIONAL)
[//]: # (Add one or multiple labels: enhancement, refactoring, bugfix)

### CODE REVIEW

File: test/writeObfuscatedMockUFC.ts [hunk 1]
```diff
@@ -11,17 +11,26 @@ import {
...
+  switch (condition.operator) {
+    case 'ONE_OF':
+    case 'NOT_ONE_OF':
+      return condition.value.map((value) => getMD5Hash(value.toLowerCase()));
+    case 'IS_NULL':
+      return getMD5Hash(`${condition.value}`);
+    default:
+      return encodeBase64(`${condition.value}`);
+  }
+}
```
Comment (schmit) [inline test/writeObfuscatedMockUFC.ts:24]: @aarsilv heads-up: adding extra logic here to handle the IS_NULL operator

File: src/evaluator.spec.ts [hunk 2]
```diff
@@ -470,6 +468,8 @@ describe('Evaluator', () => {
       totalShards: 10,
     };
 
+    console.log(flag);
```
Comment (giorgiomartini0) [inline src/evaluator.spec.ts:471]: Leftover from testing?

**Review by giorgiomartini0 — APPROVED** [review]

File: src/interfaces.ts [hunk 3]
```diff
@@ -31,9 +31,9 @@ export interface Split {
 
 export interface Allocation {
   key: string;
-  rules: Rule[];
-  startAt?: Date;
-  endAt?: Date;
+  rules?: Rule[];
+  startAt?: string; // ISO 8601
+  endAt?: string; // ISO 8601
```
Comment (aarsilv) [inline src/interfaces.ts:36]: This is annoying, but I see why we need to do this as JSON doesn't support dates.

**Review by aarsilv — APPROVED** [review]
Looking good!

File: src/interfaces.ts (hunk 3 above)
Comment (schmit) [reply src/interfaces.ts:36]: In the Python SDK, Pydantic takes care of the conversion for us. We can try to do something similar (converting to dates when parsing config) at a later time.

File: src/evaluator.spec.ts [hunk 4]
```diff
@@ -554,3 +554,101 @@ describe('Evaluator', () => {
...
+              operator: OperatorType.GTE,
+              value: 20,
+            },
+          ],
+        },
+      ];
+      const obfuscated = false;
+      expect(matchesRules(rules, { id: 'test-subject', age: 20 }, obfuscated)).toBeTruthy();
+      expect(matchesRules(rules, { id: 'test-subject', age: 10 }, obfuscated)).toBeFalsy();
+      expect(matchesRules(rules, { id: 'test-subject', country: 'UK' }, obfuscated)).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (schmit) [reply src/evaluator.spec.ts:632]: Type 'null' is not assignable to type 'AttributeType'.

[the remark this answers is not in the document]
Comment (schmit) [reply src/evaluator.spec.ts:632]: The sdk-test-data test suite has an explicit test for this though -- I think updating the type to allow for nulls is not the right move

[the remark this answers is not in the document]
Comment (aarsilv) [reply src/evaluator.spec.ts:632]: oh whoops didn't catch that--agree we can stay within the confines of typing--carry on!

[the remark this answers is not in the document]
Comment (schmit) [reply src/evaluator.spec.ts:632]: I did not catch it either -- I tried writing this test and the compiler started yelling at me

**Pull request merged**